travis: loop over apt-get to recover from errors

adrianreber opened the pull request:
One of the most common CI errors we see is that package install fails due to some hashsum mismatch or some DNS errors.

This adds a loop around each apt-get install call to do a clean, update and install and if one of the steps fails it repeats it up to 10 times.

rppt commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,35 @@ if [ "$UNAME_M" != "x86_64" ]; then
+                apt-get install -qqy --no-install-recommends $@
+                if [ $? -eq 0 ]; then
+                        break
+                fi
```
Maybe also 
```
apt-get -f install
```
just in case?

adrianreber replied:
sure

adrianreber replied:
done

rppt commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,35 @@ if [ "$UNAME_M" != "x86_64" ]; then
 fi
 
+install_retry_counter=0
+max_apt_retries=10
```
Do you think a smaller number of retries won't be sufficient?

adrianreber replied:
Initially I had 5 retries. 10 is indeed a lot and if it does not work after the third or fourth it will probably fail the tenth time.

I can change it to 5. Maybe I should add a `sleep $install_retry_counter` to give it some time to recover from errors.

adrianreber replied:
done

avagin commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,35 @@ if [ "$UNAME_M" != "x86_64" ]; then
+apt_install () {
+        set +e
+        while true; do
+                install_retry_counter=$(expr $install_retry_counter + 1)
```
let  install_retry_counter++

adrianreber replied:
`let` needs `/bin/bash`. Initially I had `let` but had to change it to `expr`. We can, of course, change the whole script to `/bin/bash`.

rppt replied:
> let install_retry_counter++

Won't work on `dash` which might be used as `/bin/sh` on Ubuntu.
Maybe we want to make travis-tests to explicitly use bash and then `let cnt++` or `((cnt++))` would work.

```
diff --git a/scripts/travis/travis-tests b/scripts/travis/travis-tests
index e9c2d5c6a..d159bb71f 100755
--- a/scripts/travis/travis-tests
+++ b/scripts/travis/travis-tests
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 set -x -e
 
 TRAVIS_PKGS="protobuf-c-compiler libprotobuf-c-dev libaio-dev python-future
```

avagin replied:
I vote to force using bash for all our shell scripts, but @adrianreber it is up to you

adrianreber replied:
switched to `let` and bash

avagin commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,35 @@ if [ "$UNAME_M" != "x86_64" ]; then
+# avoid CI errors due to errors during apt-get
+# hashsum mismatches, DNS errors and similar things
+apt_install () {
+        set +e
```
you don't need to do this

avagin commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,35 @@ if [ "$UNAME_M" != "x86_64" ]; then
+                if [ $install_retry_counter -gt $max_apt_retries ]; then
+                        exit 1
+                fi
+                apt-get clean -qqy
```
apt-get clean -qqy || continue

adrianreber commented:
Reworked and force pushed.

 * Switch to `/bin/bash` as it is probably anyway `/bin/bash` on most systems
 * Removed `set +e` and `set -e`
 * Added a `sleep` if it fails
 * Using `let` instead of `expr`

Snorch commented on scripts/travis/travis-tests:
```diff
@@ -19,6 +19,27 @@ if [ "$UNAME_M" != "x86_64" ]; then
+
+		# In case it is a network error let's wait a bit.
+		sleep $install_retry_counter
+	done
```
@adrianreber Hi, shouldn't we reset install_retry_counter to 0 at the end?

adrianreber replied:
Yes, right. Thanks. I will fix it in a follow up PR. I was planning to move it to its own script and also use it in all the Dockerfiles.
